### PR TITLE
Fix homechain ocr3

### PR DESCRIPTION
**Pull request by asoliman92**

### CODE REVIEW

File: pkg/consts/consts.go [hunk 1]
```diff
@@ -87,8 +87,9 @@ const (
 	// Used by the home chain reader.
 	MethodNameGetCapability = "GetCapability"
 
-	// CCIPConfig.sol methods.
+	// CCIPHome.sol methods.
 	// Used by the home chain reader.
+	// TODO: change them to getConfig, getAllConfigs
```
Comment (makramkd) [inline pkg/consts/consts.go:92]: This should not be a blocker for now, but lets create a followup ticket for it.

File: pkg/reader/home_chain.go [hunk 2]
```diff
@@ -19,6 +19,8 @@ type OCR3ConfigWithMeta = reader_internal.OCR3ConfigWithMeta
 
 type OCR3Config = reader_internal.OCR3Config
 
+type OCR3Node = reader_internal.OCR3Node
```
Comment (makramkd) [inline pkg/reader/home_chain.go:22]: Would love a follow up ticket to remove home chain reader from internal.

File: pkg/consts/consts.go (hunk 1 above)
Comment (asoliman92) [reply pkg/consts/consts.go:92]: https://smartcontract-it.atlassian.net/browse/CCIP-3642

File: pkg/reader/home_chain.go (hunk 2 above)
Comment (asoliman92) [reply pkg/reader/home_chain.go:22]: https://smartcontract-it.atlassian.net/browse/CCIP-3642

File: commit/chainfee/outcome.go [hunk 3]
```diff
@@ -96,7 +96,7 @@ func (p *processor) getConsensusObservation(
 ) (Observation, error) {
 	aggObs := aggregateObservations(aos)
 
-	// consensus on the fChain map uses the role DON F value
+	// consensus on the fChain map uses the role DON fRoleDon value
```
Comment (makramkd) [inline commit/chainfee/outcome.go:99]: The comment already says "the role DON F value", changing it to "the role DON fRoleDon value" seems weird.

File: commit/merkleroot/outcome.go [hunk 4]
```diff
@@ -246,7 +246,7 @@ func getConsensusObservation(
 ) (ConsensusObservation, error) {
 	aggObs := aggregateObservations(aos)
 
-	// consensus on the fChain map uses the role DON F value
+	// consensus on the fChain map uses the role DON fRoleDon value
```
Comment (makramkd) [inline commit/merkleroot/outcome.go:249]: Ditto

File: commit/tokenprice/outcome.go [hunk 5]
```diff
@@ -21,7 +21,7 @@ func (p *processor) getConsensusObservation(
 ) (ConsensusObservation, error) {
 	aggObs := aggregateObservations(aos)
 
-	// consensus on the fChain map uses the role DON F value
+	// consensus on the fChain map uses the role DON fRoleDon value
```
Comment (makramkd) [inline commit/tokenprice/outcome.go:24]: Ditto

File: go.mod [hunk 6]
```diff
@@ -5,7 +5,7 @@ go 1.22.5
 require (
 	github.com/deckarep/golang-set/v2 v2.6.0
 	github.com/smartcontractkit/chain-selectors v1.0.23
-	github.com/smartcontractkit/chainlink-common v0.2.3-0.20241001210038-dd59341432bd
+	github.com/smartcontractkit/chainlink-common v0.2.3-0.20241002211418-707d968168d4
```
Comment (makramkd) [inline go.mod:8]: What was the common bump? Unrelated?

File: internal/plugincommon/discovery/processor.go [hunk 7]
```diff
@@ -190,7 +190,7 @@ func (cdp *ContractDiscoveryProcessor) Outcome(
 	cdp.lggr.Infow("Processing contract discovery outcome", "observations", aos)
 	// come to consensus on the onramp addresses and update the chainreader.
 
-	// fChain consensus - uses the role DON F value because all nodes can observe the home chain.
+	// fChain consensus - uses the role DON fRoleDon value because all nodes can observe the home chain.
```
Comment (makramkd) [inline internal/plugincommon/discovery/processor.go:193]: Ditto

File: internal/reader/home_chain.go [hunk 8]
```diff
@@ -223,18 +223,35 @@ func (r *homeChainPoller) GetOCRConfigs(
 	ctx context.Context, donID uint32, pluginType uint8,
 ) ([]OCR3ConfigWithMeta, error) {
 	var ocrConfigs []OCR3ConfigWithMeta
+	var allConfigs GetAllConfigs
```
Comment (makramkd) [inline internal/reader/home_chain.go:226]: nit: combine into one decl

```golang
var (
  ocrConfigs []OCR3ConfigWithMeta
  allConfigs GetAllConfigs
)
```

File: internal/reader/home_chain.go [hunk 9]
```diff
@@ -223,18 +223,35 @@ func (r *homeChainPoller) GetOCRConfigs(
...
 		return nil, fmt.Errorf("error fetching OCR configs: %w", err)
 	}
 
+	r.lggr.Infow(
+		"GetOCRConfigs",
+		"activeConfig", allConfigs.ActiveConfig,
+		"candidateConfig", allConfigs.CandidateConfig,
+	)
+
+	// Not empty
```
Comment (makramkd) [inline internal/reader/home_chain.go:246]: Unnecessary comment. General comment guidelines: don't say _what_ the code is doing; if there is some obscure reasoning behind what the code is doing then include that in the comment.

File: internal/reader/home_chain.go [hunk 10]
```diff
@@ -223,18 +223,35 @@ func (r *homeChainPoller) GetOCRConfigs(
...
+	if allConfigs.ActiveConfig.Version != 0 {
+		ocrConfigs = append(ocrConfigs, allConfigs.ActiveConfig)
+	}
+	// Not empty
```
Comment (makramkd) [inline internal/reader/home_chain.go:250]: Unnecessary comment

File: internal/reader/home_chain.go [hunk 11]
```diff
@@ -348,25 +365,43 @@ type ChainConfig struct {
 	Config chainconfig.ChainConfig `json:"config"`
 }
 
-// OCR3Config mirrors CCIPConfig.sol's OCR3Config struct
+// nolint: lll
+// https://github.com/smartcontractkit/chainlink/blob/e964798a974f3246ee1da011feffe33509b358df/contracts/src/v0.8/ccip/capability/CCIPHome.sol#L105-L131
+
+type OCR3Node struct {
+	P2pID          [32]byte `json:"p2pId"`
```
Comment (makramkd) [inline internal/reader/home_chain.go:372]: This casing is just awkward IMO. I see this in many places throughout the codebase. Better:

```suggestion
	P2PID          [32]byte `json:"p2pId"`
```

File: commit/chainfee/outcome.go (hunk 3 above)
Comment (asoliman92) [reply commit/chainfee/outcome.go:99]: That was a replace all typo :D

File: go.mod (hunk 6 above)
Comment (asoliman92) [reply go.mod:8]: Was debugging locally, this is latest cl-common so it should be okay to leave it as is.

File: commit/merkleroot/types.go [hunk 12]
```diff
@@ -132,6 +132,9 @@ type Outcome struct {
 	ReportTransmissionCheckAttempts uint                          `json:"reportTransmissionCheckAttempts"`
 	RMNReportSignatures             []cciptypes.RMNECDSASignature `json:"rmnReportSignatures"`
 	RMNRemoteCfg                    rmntypes.RemoteConfig         `json:"rmnRemoteCfg"`
+	//TODO: Calculate the bitmap
+	// This is a bitmap where ith bit represents how the v value should be for ith signature
+	RmnRawVs cciptypes.BigInt `json:"rmnRawVs"`
```
Comment (mateusz-sekara) [inline commit/merkleroot/types.go:137]: nit, but all rmn related fields are capitalized (e.g. `RMNRemoteCfg`), let's keep it aligned

File: commit/merkleroot/types.go [hunk 13]
```diff
@@ -173,3 +176,23 @@ const (
...
 	WaitingForReportTransmission
 )
+
+type RootSet struct {
+	roots []cciptypes.MerkleRootChain
+}
+
+// Add a MerkleRoot to the set (if it's not already in the set)
+func (s *RootSet) Add(root cciptypes.MerkleRootChain) {
+	if !s.Contains(root) {
```
Comment (mateusz-sekara) [inline commit/merkleroot/types.go:186]: `slices.Contains()` will do exactly this, right? Also if you want to have unique values here then mapset or map would be better probably and don't care about order of insertion then mapset or map would be better choice. Insertion when calling contains on a slice gonna have O(n^2) complexity. Probably not a problem for small slices, but it is easy to forget about that when the slice grows over time.

File: internal/reader/home_chain_test.go [hunk 14]
```diff
@@ -178,18 +178,22 @@ func Test_HomeChainPoller_GetOCRConfig(t *testing.T) {
 			"donId":      donID,
 			"pluginType": pluginType,
 		},
-		mock.AnythingOfType("*[]reader.OCR3ConfigWithMeta"),
+		mock.AnythingOfType("*reader.GetAllConfigs"),
+		//mock.Anything,
```
Comment (mateusz-sekara) [inline internal/reader/home_chain_test.go:182]: remove?

File: commit/merkleroot/types.go (hunk 12 above)
Comment (makramkd) [reply commit/merkleroot/types.go]: Fixed

**Comment by makramkd** [comment]
Something weird happened to this branch, I see random commits

**Comment by makramkd** [comment]
Going to merge into fix-merge and then merge main into fix-merge. Should fix the issues hopefully.

**Pull request merged**